### DISCUSSION TITLE
World building, how close is the map creating system to being able to make game play ready maps?

### DISCUSSION THREAD

**Discussion by WithinAmnesia** [SlayHorizon/godot-tiny-mmo#31] [Q&A] [answered]
Is the map creating system ready for game play ready maps?

<details>If prototype maps are made now are they able to survive future engine iterations without needing excessive retro refitting?

I understand that demo / test maps can be made now. Yet if one person goes to make many large maps together will they be able to be used in the future of this game engine? Or will maps created now need excessive retro refitting in order to be used successfully with future game play ready iterations of this game engine?

I've got a dilemma in that I can go start on large scale world building now that I've figure out world scale and game design. Yet I risk a blind spot. In that I'm unsure if the work today will need to be thrown out or excessively retro fitted in order to work with a gameplay ready future iteration of the game engine / MMO project build.

What can be done where maps can be created for long term projects and what is needed, if anything for the project engine to be added to be ready for large scale map production of say 50+ maps with 512 x 512 tiles and with each map tile being 16 x 16 pixels in scale? How reasonable or secure is it to start work on large scale maps, even if prototyped in a pre-alpha state with rough detail silhouettes? How much work, if any would be needed for the game engine to get map creation process ready for ~40-50+ 512 x 512 tiles, with 16 x 16 pixels tiles large scale maps?

The risk is making large scale maps of ~50 maps x 512 x 512 tiles each of 16 x 16 pixels per tile = ~13,107,200+ tiles of art and then risk it having need to be redo to work with a future game play ready version / build of the Tiny MMO engine. At this grand scale of map creation even in pre-alpha with rough detail silhouettes, this is not trivial to restart or retro fit. Ideally the maps can be roughed out / worked on simultaneously with the engine work. Yet this would require reasonable map compatibility to a future engine build where possible so vast amounts of map art can be used in the future when the engine is more mature. As in with the maps being able to be used with a future build of the engine with the main game play ready features working well in with networking, combat, economy, progression, data security, user interface etc.

I wonder how close are we to having maps be able to be mass created and what work is needed to secure map compatibility with future matured engine builds? So to ensure work security in being productive today with world building / map creation that can be reasonably used in the future with a more mature game play ready engine build? I'm wondering about how future proof the map creation is currently with the engine today? Also with the pragmatic mind in thinking of what possibly could be needed to ensure that maps can be created currently that can be added to the game engine when it is ready for mass world creation? If this was a car, how close are we to the engine, drive train, controls, chassis and wheels working together to then be ready install body work, seats, furniture and fine details?

What engine features are needed for mass map creation to be reasonably future proof for full game play ready builds of the Tiny MMO engine?

All feedback is welcome and I think things can get super better, yet it's good to ask people and check in for big important steps into the future!</details>
💯 😃 👍

**Comment by WithinAmnesia** [comment]
@SlayHorizon What are your thoughts on how the public MIT example game demo for Tiny MMO Demo should be built in the future? What should this look like for general ideas? I'm curious what people in the community expect and or want for a pubic MIT demo for Tiny MMO Demo?

<details>I'm doing world building right now and I'm wondering about what demo areas I should make and also what do other people want from the public MIT example game demo for Tiny MMO Demo?

I was thinking of perhaps making 2 demos:

- A generic / Flare inspired demo 'campaign' / something like JDungeon / a small scale feature rich show case and maybe ~1-2 hours of gameplay / world scale that shows the features in the best light. With this demo being brief to be is easy to digest and full enough gameplay experience for people to want more / build their own games from being inspired with the Tiny MMO Demo. https://github.com/flareteam/flare-game + https://github.com/WithinAmnesia/ARPG/tree/main/A.R.P.G.%20Godot%204%2B/A.R.P.G.%20V.000.009%20(JDungeon)

- A shareware free version my copyright / trademarked medieval fantasy world made into an MMORPG prototype Tiny MMO Demo, with aspiring for ~1/3rd the scale of the full game as free open source share ware akin to free Doom shareware game back in the day. My game world is built to be really big and feature deep but its more restrictive for pure MIT forever free demos since it uses my own copyrights and trademarks. I could release an open source CC-BY version that is copyright to myself where people cannot sell it 'as is' with my trademarks online like with Steam, GoG, Itch.io etc but they are free to take it apart and learn and hopefully create their own game legally without my trademarks and copyrights. This option is not the best I feel for I want to lead with a MIT license free demo without legal issues for anyone and give the best template for the community to grow.

Ideally I'd like to 100% / complete the 'Generic / Flare inspired demo 'campaign' / something like JDungeon / a small scale feature rich show case of Tiny MMO Demo' first and then have that be all MIT and forever free. I would like to eventually release a shareware ~1/3rd scale demo of my much larger / grand MMORPG game / world for people to be inspired by. Yet using my own work is already trademarked so even before Tiny MMO Demo started my personal medieval fantasy setting was not MIT but I want to work on it and help this MIT MMORPG engine get to a really good start. I think the divide of 2 demos from myself at least is necessary to keep the Tiny MMO Demo all MIT license with a bundled 'Generic / Flare inspired demo 'campaign' / something like JDungeon / a small scale feature rich show case of Tiny MMO Demo'. Then I can have a discussion / a separate webpage for user demos and have my larger trademarked / copyrighted / CC-BY demo be listed as an example of a community full scale game with proper progression and world building etc.

What are your thoughts / ideas / concerns / inspirations / points on these demo creation concepts? All feedback is welcome. Things look really good too and the engine of Tiny MMO Demo is all ready impressive and I think the future can get really awesome. I'd love to get the community some more great demos to show cool world building!</details>
😃 👍 💯

**Comment by SlayHorizon** [comment, marked as answer]
If the maps are static (no networked objects), it should be fine.
You could even import maps from Tiled or other Godot projects and manually add teleporters/warps.

Currently, map object states aren’t saved or synced over the network. For example, if you break a jar, everyone nearby will see it, but if you disconnect/reconnect or someone enters the map afterward, the jar will reappear for them.

Depending on your game design, this might not be an issue—or could even be intended behavior.

Another important aspect missing for larger maps is interest management (also called zone of interest) for optimization,
but it doesn't concern map building directly.

**Reply by WithinAmnesia** [reply]
That is really good news and it helps make things more clear for future map creation and world building development.
<details>
'If the maps are static (no networked objects), it should be fine.'
- That is great to hear, I can make better prototypes and know what can be future proofed. That's really helpful knowing what is ready and can be worked on currently for demo maps / world building.

'You could even import maps from Tiled or other Godot projects and manually add teleporters/warps.'
- I really want to use Tiled.exe for I used Tiled.exe a lot with Flare and that is great to hear. I was going to ask about Tiled.exe but I was not wanting to ask too much at once lol. I'll have to get my Tiled skills up to scratch to best try to be efficient in making in maps. That's good that the teleports work and can be added by hand, I'm just happy it works and there is map to map travel.

'Currently, map object states aren’t saved or synced over the network. For example, if you break a jar, everyone nearby will see it, but if you disconnect/reconnect or someone enters the map afterward, the jar will reappear for them.

Depending on your game design, this might not be an issue—or could even be intended behavior.'
- I noticed the syncing with the demo scene as in if you open the gate then run out then run back to the first room with the gate, the gate is up even if the gate is down for the rest of the players and the button needs to be pressed again to lower the gate for the re-entered player. Also same de-sync with coins. I felt this was a non-serious issue and can be later tinkered with and improved. There are some things I feel are not big issues and some situational de-sync is okay. It feels like a patch here and there later with quality of life patches post stable test demo launch is where a lot of these minor non-breaking bugs can be improved / fixed bit by bit over time. I'm glad they are not too serious and can be addressed later and it helps figure out whats needed to work now for map features and what can be later addressed with future tinkering. The gate can be made with switches on both sides and the coins I know of a syncing multiplayer code that could help from: https://github.com/Zylann/voxelgame + https://github.com/WithinAmnesia/ARPG/discussions/16#discussioncomment-8513009 + https://github.com/WithinAmnesia/ARPG/discussions/16#discussioncomment-8676468

'Another important aspect missing for larger maps is interest management (also called zone of interest) for optimization,
but it doesn't concern map building directly.'
- Yeah I agree, especially in that really good games have beautiful maps made with soul and they just feel full of life. It's a game design thing mostly, but yeah having big empty maps is a technical feature but details and good quality environments can be added to empty maps bit by bit. Yet it needs to have a big map first in order to ensure the future proofing of layer by layer map detail over time. Just like WoW vanilla world map development from alpha to launch: https://old.wow.tools/maps/Azeroth/4/4/-96.125/62.688 It took from ~1999-2004 to get ~3/4 of the world map complete ready for gameplay with a full art map making dev team as well. Yet WoW started with big open areas / maps that were bit by bit handcrafted to add in many cool points of interest.

This is all really good to hear and I feel I can start making static maps with the fundamental map features / points of interest silhouettes details and then properly add teleportation / map to map travel and start making prototypes for the demes world building. I am thrilled to hear that tiled.exe can be used right away for the map making process as well, this makes things really awesome to make maps with. I wonder what the demos will look like roughed out, only one way to find out and start more map making game development. I should get more updates on these map making world building for the new Demos as time progresses. I should assemble the tilesets and map assets for MIT art and work on the MIT smaller scale demo and tinker more with Tiled.exe to get the maps working will for Tiny MMO Demo. Once I get this figured out I should be able to help write and document with screenshots the wiki and documentation for people in the community a how to guide on making maps for world building with Tiny MMO Demo. What kind of examples / feature show case / what cool things / locations should be in the public MIT ~1-2 hour long game play campaign demo for Tiny MMO Demo? This project is going really awesome so far. All ideas, concepts, thoughts, questions, curiosities and feedback are welcome!</details>
😃 👍 ✨ 💯 🏰 🌎

**Reply by WithinAmnesia** [reply]
![M M O A R P G (V 000 075) Godot-Tiny-MMO-Demo-0 0 29 Godot 4 3 NET+ Godot Tiny MMO (Demo 1 Overworld!)](https://github.com/user-attachments/assets/badff809-fd1c-4a97-97fe-e00a417e88e8)
My first world building Demo map working for ~'MIT Main Public Demo Showcase Campaign' for Tiny MMO Demo!
<details>
I added another test map, its themed after the dungeon entrance into the wider world map; so I called the new area 'WorldMap' and I got the map to map travel teleporting / warping in multiplayer working and I also got the a new gate spikes location working and I figured out the collision and how to tinker and world build as is in the game engine for the  ~'MIT Main Public Demo Showcase Campaign' for Tiny MMO Demo.

I can start experimenting / tinkering with world and map scale for some rough stress test with ~16+ players and begin to work on my CC-BY game world too with a second demo for my own works in due time. The ~'MIT Main Public Demo Showcase Campaign' for Tiny MMO Demo should take some priority of my focus to get working well enough as well to help the community. This is a really great benchmark for making more maps. I could polish this up later and perhaps share a better ~'MIT Main Public Demo Showcase Campaign' for Tiny MMO Demo later for better community support when its ready.

I think combat / abilities / spells and mobs and entities are needed for proper full scale map stress testing for the next big progress benchmark for world building with Tiny MMO Demo. I wonder when this can happen realistically? I'd like to help add combat and entities if really needed but I know I'm not always the best coder and I could bungle things vs if I let time naturally flow for better code to have the chance to be written first; so the combat and entities are written with really solid / clean code vs my spaghetti monster code lol. Also want to give lots of freedom and thinking space for open source dev work (good work environment and high moral are really important for the community in general) but if there are problems with entities / mobs this version of JDungeon MMORPG Demo I archived has some code that might help for combat / abilities and entities: https://github.com/WithinAmnesia/ARPG/tree/main/A.R.P.G.%20Godot%204%2B/A.R.P.G.%20V.000.009%20(JDungeon)

This is really good progress so far! What are your thoughts / feedback/ questions? All feedback is welcome.
</details>
🌎 🌍 🌏 💪 😃 👍 ✨ 🏰 💯

**Reply by SlayHorizon** [reply]
I can’t see much detail, but with so many game instances running, isn’t that making your PC lag?

Regarding the combat system, another contributor is currently working on a new animation system that will support various graphics and weapon animations. I’m waiting for that to be completed before finalizing the combat system. However, the character system already allows you to create some NPCs, though there isn’t yet a proper way to test them.

**Reply by WithinAmnesia** [reply]
https://github.com/SlayHorizon/godot-tiny-mmo-demo/discussions/31#discussioncomment-10939401
That's really great to hear. I was away yesterday for Canadian Thanksgiving so I just got back now. That is super excellent / awesome to hear about the combat. I was tempted to look at the JDungeon and Blockygame combat systems for inspiration lol.

<details>
The picture is a rough or like 1st time having it all working. My PC does not lag with so much for it was built to handle 4k unoptimized game dev (Unity and Unreal 4k to 8k). So it's good for testing but I do have a regular steam machine 1080p kinda laptop to test for optimization for most all of the community to enjoy. The screenshot is sorta like my 1st time getting it all working with a new map / testing the world building / map making. Its sorta like a benchmark in time.

"Another contributor is currently working on a new animation system that will support various graphics and weapon animations" Is this the 2 direction animation or is it the 4 directional animation? Is it basically this?
![0vizkyr](https://github.com/user-attachments/assets/b8e01fb4-372e-490f-b959-ef847a9274d5)
Or is it more like ARPG like Diablo / Flare / Path of Exile / Zelda?
![Warrior Showreel](https://github.com/user-attachments/assets/cf37ed61-f452-4314-9d4a-3a6a65db5c08)
Is it like this 4-directional animation system or I wonder if both types can be supported together? That would be the best of both worlds.


"However, the character system already allows you to create some NPCs, though there isn’t yet a proper way to test them."
Is this meaning like one can art swap the character animation art on the characters to test out the animations? Are we at the point where mobs / entities can be placed in the world scene / server and move around / be interacted with? I wonder what NPC options there are to test and explore? I'll have to check out the new version more and tinker around. It looks really good and if we get combat, entities and progression systems rocking and rolling this Tiny MMO Demo is basically super ripped / strong. This can super help the creative community so much get their boots planted and help creative people get their feet on the ground and running full steam ahead onwards. Life is much more fun when there is a big adventure.
</details>
💪 😃 👍 ✨ 💯 🗺️
